File: cpp/gql/src/libgql/src/libgql/parsers/schema/visitor.cpp
#include "./visitor.hpp"

#include <memory>
#include <ranges>
#include <variant>

#include "./schema.hpp"
#include "libgql/parsers/schema/client_ast.hpp"
#include "libgql/parsers/schema/server_ast.hpp"
#include "libgql/parsers/schema/shared_ast.hpp"
#include "utils.hpp"

namespace gql::parsers::schema::visitor {
void visitFieldSelection(const ASTVisitorHooks &hooks,
                         const ast::FieldSelection &selection) {
    if (hooks.visitFieldSelection.has_value()) {
        hooks.visitFieldSelection.value()(selection);
    };
    for (const auto &argument : selection.arguments | std::views::values) {
        if (hooks.visitFieldSelectionArgument.has_value()) {
            hooks.visitFieldSelectionArgument.value()(argument);
        };
        if (hooks.visitArgumentValue.has_value()) {
            hooks.visitArgumentValue.value()(argument.value);
        };
        std::visit(
            utils::overloaded{
                [&hooks](const ast::ArgumentRefValue &value) -> void {
                    if (hooks.visitArgumentRefValue.has_value()) {
                        hooks.visitArgumentRefValue.value()(value);
                    };
                },
                [&hooks](const ast::ArgumentLiteralValue &value) -> void {
                    if (hooks.visitArgumentLiteralValue.has_value()) {
                        hooks.visitArgumentLiteralValue.value()(value);
                    };
                },
            },
            argument.value);
    };
    if (selection.selection.has_value()) {
        const auto &fSelection = *selection.selection.value().get();
        visitFragmentSpec(hooks, fSelection);
    };
};

void visitObjectFieldSpec(const ASTVisitorHooks &hooks,
                          const ast::ObjectFieldSpec &spec);
void visitFieldDefinition_ObjectFieldSpec(
    const ASTVisitorHooks &hooks,
    const std::shared_ptr<ast::FieldDefinition<ast::ObjectFieldSpec>> &field) {
    if (hooks.visitFieldDefinition_ObjectFieldSpec.has_value()) {
        hooks.visitFieldDefinition_ObjectFieldSpec.value()(field);
    };
    visitObjectFieldSpec(hooks, field->spec);
};

template <typename T>
void visitObjectSelection(const ASTVisitorHooks &hooks,
                          const ast::ObjectSelection &objectSelection,
                          const std::shared_ptr<T> &object) {
    if (hooks.visitObjectSelection.has_value()) {
        hooks.visitObjectSelection.value()(objectSelection);
    };
    std::visit(utils::overloaded{
                   [&hooks](const ast::TypenameField &field) -> void {
                       if (hooks.visitTypenameField.has_value()) {
                           hooks.visitTypenameField.value()(field);
                       };
                   },
                   [&hooks](const ast::SpreadSelection &field) -> void {
                       if (hooks.visitSpreadSelection.has_value()) {
                           hooks.visitSpreadSelection.value()(field);
                       };
                       if (hooks.visitFragment.has_value()) {
                           hooks.visitFragment.value()(field.fragment);
                       };
                       visitFragmentSpec(hooks, field.fragment->spec);
                   },
                   [&hooks, &object](const ast::FieldSelection &field) -> void {
                       visitFieldSelection(hooks, field);
                       const auto &f = object->fields[field.name];
                       visitFieldDefinition_ObjectFieldSpec(hooks, f);
                   },
               },
               objectSelection);
};

void visitObjectFragmentSpec_ObjectType(
    const ASTVisitorHooks &hooks,
    const ast::ObjectFragmentSpec<ast::ObjectType> &spec);
void visitUnionSelection(const ASTVisitorHooks &hooks,
                         const ast::UnionSelection &selection) {
    if (hooks.visitUnionSelection.has_value()) {
        hooks.visitUnionSelection.value()(selection);
    };
    std::visit(
        utils::overloaded{
            [&hooks](const ast::TypenameField &field) -> void {
                if (hooks.visitTypenameField.has_value()) {
                    hooks.visitTypenameField.value()(field);
                };
            },
            [&hooks](const ast::SpreadSelection &field) -> void {
                if (hooks.visitSpreadSelection.has_value()) {
                    hooks.visitSpreadSelection.value()(field);
                };
                if (hooks.visitFragment.has_value()) {
                    hooks.visitFragment.value()(field.fragment);
                };
                visitFragmentSpec(hooks, field.fragment->spec);
            },
            [&hooks](
                const ast::ObjectConditionalSpreadSelection &field) -> void {
                if (hooks.visitObjectConditionalSpreadSelection.has_value()) {
                    hooks.visitObjectConditionalSpreadSelection.value()(field);
                };
                visitObjectFragmentSpec_ObjectType(hooks,
                                                   *field.selection.get());
            },
            [](const ast::UnionConditionalSpreadSelection &) {} },
        selection);
};

void visitObjectFragmentSpec_ObjectType(
    const ASTVisitorHooks &hooks,
    const ast::ObjectFragmentSpec<ast::ObjectType> &spec) {
    if (hooks.visitObjectFragmentSpec_ObjectType.has_value()) {
        hooks.visitObjectFragmentSpec_ObjectType.value()(spec);
    };
    if (hooks.visitObjectType.has_value()) {
        hooks.visitObjectType.value()(spec.type);
    };
    for (const auto &selection : spec.selections) {
        visitObjectSelection(hooks, selection, spec.type);
    };
};

void visitObjectFragmentSpec_Interface(
    const ASTVisitorHooks &hooks,
    const ast::ObjectFragmentSpec<ast::Interface> &spec) {
    if (hooks.visitObjectFragmentSpec_Interface.has_value()) {
        hooks.visitObjectFragmentSpec_Interface.value()(spec);
    };
    for (const auto &selection : spec.selections) {
        visitObjectSelection(hooks, selection, spec.type);
    };
};

void visitFieldDefinition_InputFieldSpec(
    const ASTVisitorHooks &hooks,
    const ast::FieldDefinition<ast::InputFieldSpec> &field);
void visitInputType(const ASTVisitorHooks &hooks,
                    const std::shared_ptr<ast::InputType> &input) {
    if (hooks.visitInputType.has_value()) {
        hooks.visitInputType.value()(input);
    };
    for (const auto &field : input->fields | std::views::values) {
        visitFieldDefinition_InputFieldSpec(hooks, field);
    };
};

void visitInputTypeSpec(const ASTVisitorHooks &hooks,
                        const ast::InputTypeSpec &spec) {
    if (hooks.visitInputTypeSpec.has_value()) {
        hooks.visitInputTypeSpec.value()(spec);
    };
    std::visit(utils::overloaded{
                   [&hooks](const std::shared_ptr<ast::Scalar> &scalar) {
                       if (hooks.visitScalar.has_value()) {
                           hooks.visitScalar.value()(scalar);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::Enum> &enumType) {
                       if (hooks.visitEnum.has_value()) {
                           hooks.visitEnum.value()(enumType);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::InputType> &input) {
                       visitInputType(hooks, input);
                   } },
               spec);
};

void visitObjectTypeSpec(const ASTVisitorHooks &hooks,
                         const ast::ObjectTypeSpec &spec) {
    if (hooks.visitObjectTypeSpec.has_value()) {
        hooks.visitObjectTypeSpec.value()(spec);
    };
    std::visit(utils::overloaded{
                   [&hooks](const std::shared_ptr<ast::Scalar> &scalar) {
                       if (hooks.visitScalar.has_value()) {
                           hooks.visitScalar.value()(scalar);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::Enum> &enumType) {
                       if (hooks.visitEnum.has_value()) {
                           hooks.visitEnum.value()(enumType);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::ObjectType> &object) {
                       if (hooks.visitObjectType.has_value()) {
                           hooks.visitObjectType.value()(object);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::Interface> &interface) {
                       if (hooks.visitInterface.has_value()) {
                           hooks.visitInterface.value()(interface);
                       };
                   },
                   [&hooks](const std::shared_ptr<ast::Union> &unionType) {
                       if (hooks.visitUnion.has_value()) {
                           hooks.visitUnion.value()(unionType);
                       };
                   } },
               spec);
};

void visitInputFieldSpec(const ASTVisitorHooks &hooks,
                         const ast::InputFieldSpec &spec) {
    if (hooks.visitNonCallableFieldSpec_InputTypeSpec.has_value()) {
        hooks.visitNonCallableFieldSpec_InputTypeSpec.value()(spec);
    };
    std::visit(
        utils::overloaded{
            [&hooks](const ast::LiteralFieldSpec<ast::InputTypeSpec> &literal) {
                if (hooks.visitLiteralFieldSpec_InputTypeSpec.has_value()) {
                    hooks.visitLiteralFieldSpec_InputTypeSpec.value()(literal);
                };
                visitInputTypeSpec(hooks, literal.type);
            },
            [&hooks](const ast::ArrayFieldSpec<ast::InputTypeSpec> &array) {
                if (hooks.visitArrayFieldSpec_InputTypeSpec.has_value()) {
                    hooks.visitArrayFieldSpec_InputTypeSpec.value()(array);
                };
                visitInputTypeSpec(hooks, array.type);
            },
        },
        spec);
};

void visitNonCallableFieldSpec_ObjectTypeSpec(
    const ASTVisitorHooks &hooks,
    const ast::NonCallableFieldSpec<ast::ObjectTypeSpec> &spec) {
    if (hooks.visitNonCallableFieldSpec_ObjectTypeSpec.has_value()) {
        hooks.visitNonCallableFieldSpec_ObjectTypeSpec.value()(spec);
    };
    std::visit(
        utils::overloaded{
            [&hooks](
                const ast::LiteralFieldSpec<ast::ObjectTypeSpec> &literal) {
                if (hooks.visitLiteralFieldSpec_ObjectTypeSpec.has_value()) {
                    hooks.visitLiteralFieldSpec_ObjectTypeSpec.value()(literal);
                };
                visitObjectTypeSpec(hooks, literal.type);
            },
            [&hooks](const ast::ArrayFieldSpec<ast::ObjectTypeSpec> &array) {
                if (hooks.visitArrayFieldSpec_ObjectTypeSpec.has_value()) {
                    hooks.visitArrayFieldSpec_ObjectTypeSpec.value()(array);
                };
                visitObjectTypeSpec(hooks, array.type);
            },
        },
        spec);
};

void visitObjectFieldSpec(const ASTVisitorHooks &hooks,
                          const ast::ObjectFieldSpec &spec) {
    if (hooks.visitObjectFieldSpec.has_value()) {
        hooks.visitObjectFieldSpec.value()(spec);
    };
    std::visit(
        utils::overloaded{
            [&hooks](
                const ast::LiteralFieldSpec<ast::ObjectTypeSpec> &literal) {
                if (hooks.visitLiteralFieldSpec_ObjectTypeSpec.has_value()) {
                    hooks.visitLiteralFieldSpec_ObjectTypeSpec.value()(literal);
                };
                visitObjectTypeSpec(hooks, literal.type);
            },
            [&hooks](const ast::ArrayFieldSpec<ast::ObjectTypeSpec> &array) {
                if (hooks.visitArrayFieldSpec_ObjectTypeSpec.has_value()) {
                    hooks.visitArrayFieldSpec_ObjectTypeSpec.value()(array);
                };
                visitObjectTypeSpec(hooks, array.type);
            },
            [&hooks](const ast::CallableFieldSpec &spec) {
                if (hooks.visitCallableFieldSpec.has_value()) {
                    hooks.visitCallableFieldSpec.value()(spec);
                };
                for (const auto &argument :
                     spec.arguments | std::views::values) {
                    visitFieldDefinition_InputFieldSpec(hooks, *argument.get());
                };
                visitNonCallableFieldSpec_ObjectTypeSpec(hooks,
                                                         spec.returnType);
            },
        },
        spec);
};

void visitFieldDefinition_InputFieldSpec(
    const ASTVisitorHooks &hooks,
    const ast::FieldDefinition<ast::InputFieldSpec> &field) {
    if (hooks.visitFieldDefinition_InputFieldSpec.has_value()) {
        hooks.visitFieldDefinition_InputFieldSpec.value()(field);
    };
    visitInputFieldSpec(hooks, field.spec);
};

void visitUnion(const ASTVisitorHooks &hooks,
                const std::shared_ptr<ast::Union> &unionType) {
    if (hooks.visitUnion.has_value()) {
        hooks.visitUnion.value()(unionType);
    };
    for (const auto &object : unionType->items | std::views::values) {
        if (hooks.visitObjectType.has_value()) {
            hooks.visitObjectType.value()(object);
        };
        for (const auto &field : object->fields | std::views::values) {
            visitObjectFieldSpec(hooks, field->spec);
        };
    };
};

void visitFragmentSpec(const ASTVisitorHooks &hooks,
                       const ast::FragmentSpec &fragmentSpec) {
    if (hooks.visitFragmentSpec.has_value()) {
        hooks.visitFragmentSpec.value()(fragmentSpec);
    };
    std::visit(
        utils::overloaded{
            [&hooks](const ast::ObjectFragmentSpec<ast::ObjectType> &spec) {
                visitObjectFragmentSpec_ObjectType(hooks, spec);
            },
            [&hooks](const ast::ObjectFragmentSpec<ast::Interface> &spec) {
                visitObjectFragmentSpec_Interface(hooks, spec);
            },
            [&hooks](const ast::UnionFragmentSpec &spec) {
                if (hooks.visitUnionFragmentSpec.has_value()) {
                    hooks.visitUnionFragmentSpec.value()(spec);
                };
                visitUnion(hooks, spec.type);
                for (const auto &selection : spec.selections) {
                    visitUnionSelection(hooks, selection);
                };
            } },
        fragmentSpec);
};

void visitClientSchema(const ASTVisitorHooks &hooks,
                       const ClientSchema &schema) {
    for (const auto &directive : schema.directives | std::views::values) {
        if (hooks.visitClientDirective.has_value()) {
            hooks.visitClientDirective.value()(directive);
        };
        if (hooks.visitClientDirectiveLocation.has_value()) {
            for (const auto &location : directive->locations) {
                hooks.visitClientDirectiveLocation.value()(location);
            };
        };
    };
    for (const auto &fragment : schema.fragments | std::views::values) {
        if (hooks.visitFragment.has_value()) {
            hooks.visitFragment.value()(fragment);
        };
        visitFragmentSpec(hooks, fragment->spec);
    };

    for (const auto &operation : schema.operations | std::views::values) {
        if (hooks.visitOperation.has_value()) {
            hooks.visitOperation.value()(operation);
        };
        for (const auto &parameter :
             operation->parameters | std::views::values) {
            visitFieldDefinition_InputFieldSpec(hooks, parameter);
        };
        visitFragmentSpec(hooks, operation->fragmentSpec);
    };
};

void visitServerSchema(const ASTVisitorHooks &hooks,
                       const ServerSchema &schema) {};
};  // namespace gql::parsers::schema::visitor

File: rust/gql/libgql/src/parsers/schema/visitor.rs
use std::{cell::RefCell, rc::Rc};

use crate::parsers::schema::{client, server, shared};

pub type ASTVisitorHook<'a, T> = Option<Box<dyn FnMut(&T) + 'a>>;

#[derive(Default)]
pub struct ASTVisitorHooks<'a> {
    pub visit_fragment: ASTVisitorHook<'a, Rc<RefCell<client::ast::Fragment>>>,
    pub visit_fragment_spec: ASTVisitorHook<'a, client::ast::FragmentSpec>,
    pub visit_union_fragment_spec:
        ASTVisitorHook<'a, client::ast::UnionFragmentSpec>,
    pub visit_union_selection: ASTVisitorHook<'a, client::ast::UnionSelection>,
    pub visit_typename_field: ASTVisitorHook<'a, client::ast::TypenameField>,
    pub visit_spread_selection:
        ASTVisitorHook<'a, client::ast::SpreadSelection>,
    pub visit_object_conditional_spread_selection:
        ASTVisitorHook<'a, client::ast::ObjectConditionalSpreadSelection>,
    pub visit_object_type:
        ASTVisitorHook<'a, Rc<RefCell<server::ast::ObjectType>>>,
    pub visit_object_fragment_spec_object_type: ASTVisitorHook<
        'a,
        client::ast::ObjectFragmentSpec<server::ast::ObjectType>,
    >,
    pub visit_object_fragment_spec_interface: ASTVisitorHook<
        'a,
        client::ast::ObjectFragmentSpec<server::ast::Interface>,
    >,
    pub visit_interface:
        ASTVisitorHook<'a, Rc<RefCell<server::ast::Interface>>>,
    pub visit_object_selection:
        ASTVisitorHook<'a, client::ast::ObjectSelection>,
    pub visit_field_selection: ASTVisitorHook<'a, client::ast::FieldSelection>,
    pub visit_field_selection_argument:
        ASTVisitorHook<'a, shared::ast::FieldSelectionArgument>,
    pub visit_argument_value: ASTVisitorHook<'a, shared::ast::ArgumentValue>,
    pub visit_argument_ref_value: ASTVisitorHook<'a, String>,
    pub visit_argument_literal_value:
        ASTVisitorHook<'a, shared::ast::ArgumentLiteralValue>,
    pub visit_client_directive:
        ASTVisitorHook<'a, Rc<client::ast::ClientDirective>>,
    pub visit_client_directive_location:
        ASTVisitorHook<'a, client::ast::DirectiveLocation>,
    pub visit_field_definition_input_field_spec: ASTVisitorHook<
        'a,
        shared::ast::FieldDefinition<shared::ast::InputFieldSpec>,
    >,
    pub visit_field_definition_object_field_spec: ASTVisitorHook<
        'a,
        shared::ast::FieldDefinition<server::ast::ObjectFieldSpec>,
    >,
    pub visit_non_callable_field_spec_input_type_spec: ASTVisitorHook<
        'a,
        shared::ast::NonCallableFieldSpec<shared::ast::InputTypeSpec>,
    >,
    pub visit_literal_field_spec_input_type_spec: ASTVisitorHook<
        'a,
        shared::ast::LiteralFieldSpec<shared::ast::InputTypeSpec>,
    >,
    pub visit_array_field_spec_input_type_spec: ASTVisitorHook<
        'a,
        shared::ast::ArrayFieldSpec<shared::ast::InputTypeSpec>,
    >,
    pub visit_input_type_spec: ASTVisitorHook<'a, shared::ast::InputTypeSpec>,
    pub visit_non_callable_field_spec_object_type_spec: ASTVisitorHook<
        'a,
        shared::ast::NonCallableFieldSpec<server::ast::ObjectTypeSpec>,
    >,
    pub visit_literal_field_spec_object_type_spec: ASTVisitorHook<
        'a,
        shared::ast::LiteralFieldSpec<server::ast::ObjectTypeSpec>,
    >,
    pub visit_array_field_spec_object_type_spec: ASTVisitorHook<
        'a,
        shared::ast::ArrayFieldSpec<server::ast::ObjectTypeSpec>,
    >,
    pub visit_object_field_spec:
        ASTVisitorHook<'a, server::ast::ObjectFieldSpec>,
    pub visit_callable_field_spec:
        ASTVisitorHook<'a, server::ast::CallableFieldSpec>,
    pub visit_object_type_spec: ASTVisitorHook<'a, server::ast::ObjectTypeSpec>,
    pub visit_operation:
        ASTVisitorHook<'a, Rc<RefCell<client::ast::Operation>>>,
    pub visit_input_type:
        ASTVisitorHook<'a, Rc<RefCell<shared::ast::InputType>>>,
    pub visit_scalar: ASTVisitorHook<'a, String>,
    pub visit_enum: ASTVisitorHook<'a, Rc<RefCell<shared::ast::Enum>>>,
    pub visit_union: ASTVisitorHook<'a, Rc<RefCell<server::ast::Union>>>,
}

fn visit_field_selection(
    hooks: &mut ASTVisitorHooks,
    selection: &client::ast::FieldSelection,
) {
    if let Some(hook) = hooks.visit_field_selection.as_mut() {
        hook(selection);
    }
    for argument in selection.arguments.values() {
        if let Some(hook) = hooks.visit_field_selection_argument.as_mut() {
            hook(argument);
        }
        if let Some(hook) = hooks.visit_argument_value.as_mut() {
            hook(&argument.value);
        }
        match &argument.value {
            shared::ast::ArgumentValue::Ref(r) => {
                if let Some(hook) = hooks.visit_argument_ref_value.as_mut() {
                    hook(r);
                }
            }
            shared::ast::ArgumentValue::Literal(literal) => {
                if let Some(hook) = hooks.visit_argument_literal_value.as_mut()
                {
                    hook(literal);
                }
            }
        }
    }
    if let Some(s) = &selection.selection {
        visit_fragment_spec(hooks, s);
    }
}

fn visit_field_definition_object_field_spec(
    hooks: &mut ASTVisitorHooks,
    field: &shared::ast::FieldDefinition<server::ast::ObjectFieldSpec>,
) {
    if let Some(hook) = hooks.visit_field_definition_object_field_spec.as_mut()
    {
        hook(field);
    }
    visit_object_field_spec(hooks, &field.spec)
}

fn visit_object_selection(
    hooks: &mut ASTVisitorHooks,
    selection: &client::ast::ObjectSelection,
    fields: &indexmap::IndexMap<
        String,
        Rc<shared::ast::FieldDefinition<server::ast::ObjectFieldSpec>>,
    >,
) {
    if let Some(hook) = hooks.visit_object_selection.as_mut() {
        hook(selection);
    }
    match selection {
        client::ast::ObjectSelection::TypenameField(typename_field) => {
            if let Some(hook) = hooks.visit_typename_field.as_mut() {
                hook(typename_field);
            }
        }
        client::ast::ObjectSelection::FieldSelection(field) => {
            visit_field_selection(hooks, field);
            let f = fields.get(&field.name).unwrap();
            visit_field_definition_object_field_spec(hooks, f);
        }
        client::ast::ObjectSelection::SpreadSelection(spread) => {
            if let Some(hook) = hooks.visit_spread_selection.as_mut() {
                hook(spread);
            }
            if let Some(hook) = hooks.visit_fragment.as_mut() {
                hook(&spread.fragment)
            }
            visit_fragment_spec(hooks, &spread.fragment.borrow().spec)
        }
    }
}

fn visit_object_fragment_spec_object_type(
    hooks: &mut ASTVisitorHooks,
    spec: &client::ast::ObjectFragmentSpec<server::ast::ObjectType>,
) {
    if let Some(hook) = hooks.visit_object_fragment_spec_object_type.as_mut() {
        hook(spec);
    }
    if let Some(hook) = hooks.visit_object_type.as_mut() {
        hook(&spec.r#type);
    }
    for selection in spec.selections.iter() {
        visit_object_selection(hooks, selection, &spec.r#type.borrow().fields)
    }
}

fn visit_object_type_spec(
    hooks: &mut ASTVisitorHooks,
    type_spec: &server::ast::ObjectTypeSpec,
) {
    if let Some(hook) = hooks.visit_object_type_spec.as_mut() {
        hook(type_spec);
    }
    match type_spec {
        server::ast::ObjectTypeSpec::Interface(interface) => {
            if let Some(hook) = hooks.visit_interface.as_mut() {
                hook(interface);
            }
        }
        server::ast::ObjectTypeSpec::Enum(e) => {
            if let Some(hook) = hooks.visit_enum.as_mut() {
                hook(e);
            }
        }
        server::ast::ObjectTypeSpec::ObjectType(o) => {
            if let Some(hook) = hooks.visit_object_type.as_mut() {
                hook(o);
            }
        }
        server::ast::ObjectTypeSpec::Scalar { name } => {
            if let Some(hook) = hooks.visit_scalar.as_mut() {
                hook(name);
            }
        }
        server::ast::ObjectTypeSpec::Union(union) => {
            if let Some(hook) = hooks.visit_union.as_mut() {
                hook(union);
            }
        }
    }
}

fn visit_object_fragment_spec_interface_type(
    hooks: &mut ASTVisitorHooks,
    spec: &client::ast::ObjectFragmentSpec<server::ast::Interface>,
) {
    if let Some(hook) = hooks.visit_object_fragment_spec_interface.as_mut() {
        hook(spec);
    }
    for selection in spec.selections.iter() {
        visit_object_selection(hooks, selection, &spec.r#type.borrow().fields)
    }
}

fn visit_non_callable_field_spec_object_type_spec(
    hooks: &mut ASTVisitorHooks,
    field: &shared::ast::NonCallableFieldSpec<server::ast::ObjectTypeSpec>,
) {
    if let Some(hook) = hooks
        .visit_non_callable_field_spec_object_type_spec
        .as_mut()
    {
        hook(field);
    }
    match field {
        shared::ast::NonCallableFieldSpec::Literal(literal) => {
            if let Some(hook) =
                hooks.visit_literal_field_spec_object_type_spec.as_mut()
            {
                hook(literal);
            }
            visit_object_type_spec(hooks, &literal.r#type);
        }
        shared::ast::NonCallableFieldSpec::Array(array) => {
            if let Some(hook) =
                hooks.visit_array_field_spec_object_type_spec.as_mut()
            {
                hook(array);
            }
            visit_object_type_spec(hooks, &array.r#type);
        }
    }
}

fn visit_object_field_spec(
    hooks: &mut ASTVisitorHooks,
    spec: &server::ast::ObjectFieldSpec,
) {
    if let Some(hook) = hooks.visit_object_field_spec.as_mut() {
        hook(spec);
    }
    match spec {
        server::ast::ObjectFieldSpec::Literal(literal) => {
            if let Some(hook) =
                hooks.visit_literal_field_spec_object_type_spec.as_mut()
            {
                hook(literal);
            }
            visit_object_type_spec(hooks, &literal.r#type);
        }
        server::ast::ObjectFieldSpec::Array(array) => {
            if let Some(hook) =
                hooks.visit_array_field_spec_object_type_spec.as_mut()
            {
                hook(array);
            }
            visit_object_type_spec(hooks, &array.r#type);
        }
        server::ast::ObjectFieldSpec::Callable(callable) => {
            if let Some(hook) = hooks.visit_callable_field_spec.as_mut() {
                hook(callable);
            }
            for argument in callable.arguments.values() {
                visit_field_definition_input_field_spec(hooks, argument);
            }
            visit_non_callable_field_spec_object_type_spec(
                hooks,
                &callable.return_type,
            );
        }
    }
}

fn visit_union(
    hooks: &mut ASTVisitorHooks,
    union: &Rc<RefCell<server::ast::Union>>,
) {
    if let Some(hook) = hooks.visit_union.as_mut() {
        hook(union);
    }
    for object in union.borrow().items.values() {
        if let Some(hook) = hooks.visit_object_type.as_mut() {
            hook(object);
        }
        for field in object.borrow().fields.values() {
            visit_object_field_spec(hooks, &field.spec);
        }
    }
}

fn visit_union_selection(
    hooks: &mut ASTVisitorHooks,
    selection: &client::ast::UnionSelection,
) {
    if let Some(hook) = hooks.visit_union_selection.as_mut() {
        hook(selection);
    }
    match selection {
        client::ast::UnionSelection::TypenameField(field) => {
            if let Some(hook) = hooks.visit_typename_field.as_mut() {
                hook(field);
            }
        }
        client::ast::UnionSelection::SpreadSelection(selection) => {
            if let Some(hook) = hooks.visit_spread_selection.as_mut() {
                hook(selection);
            }
            if let Some(hook) = hooks.visit_fragment.as_mut() {
                hook(&selection.fragment)
            }
            visit_fragment_spec(hooks, &selection.fragment.borrow().spec)
        }
        client::ast::UnionSelection::UnionConditionalSpreadSelection(_) => {}
        client::ast::UnionSelection::ObjectConditionalSpreadSelection(
            selection,
        ) => {
            if let Some(hook) =
                hooks.visit_object_conditional_spread_selection.as_mut()
            {
                hook(selection);
            }
            visit_object_fragment_spec_object_type(hooks, &selection.selection)
        }
    }
}

fn visit_union_fragment_spec(
    hooks: &mut ASTVisitorHooks,
    spec: &client::ast::UnionFragmentSpec,
) {
    if let Some(hook) = hooks.visit_union_fragment_spec.as_mut() {
        hook(spec);
    };
    visit_union(hooks, &spec.r#type);
    for selection in spec.selections.iter() {
        visit_union_selection(hooks, selection);
    }
}

fn visit_fragment_spec(
    hooks: &mut ASTVisitorHooks,
    fragment_spec: &client::ast::FragmentSpec,
) {
    if let Some(hook) = hooks.visit_fragment_spec.as_mut() {
        hook(fragment_spec);
    }
    match fragment_spec {
        client::ast::FragmentSpec::Object(v) => {
            visit_object_fragment_spec_object_type(hooks, v);
        }
        client::ast::FragmentSpec::Interface(v) => {
            visit_object_fragment_spec_interface_type(hooks, v);
        }
        client::ast::FragmentSpec::Union(v) => {
            visit_union_fragment_spec(hooks, v);
        }
    };
}

fn visit_input_type(
    hooks: &mut ASTVisitorHooks,
    t: &Rc<RefCell<shared::ast::InputType>>,
) {
    if let Some(hook) = hooks.visit_input_type.as_mut() {
        hook(t)
    }
    for field in t.borrow().fields.values() {
        visit_field_definition_input_field_spec(hooks, field);
    }
}

fn visit_input_type_spec(
    hooks: &mut ASTVisitorHooks,
    spec: &shared::ast::InputTypeSpec,
) {
    if let Some(hook) = hooks.visit_input_type_spec.as_mut() {
        hook(spec)
    }
    match spec {
        shared::ast::InputTypeSpec::Enum(e) => {
            if let Some(hook) = hooks.visit_enum.as_mut() {
                hook(e)
            }
        }
        shared::ast::InputTypeSpec::Scalar(s) => {
            if let Some(hook) = hooks.visit_scalar.as_mut() {
                hook(s)
            }
        }
        shared::ast::InputTypeSpec::InputType(input) => {
            visit_input_type(hooks, input);
        }
    }
}

fn visit_input_field_spec(
    hooks: &mut ASTVisitorHooks,
    field_spec: &shared::ast::InputFieldSpec,
) {
    if let Some(hook) =
        hooks.visit_non_callable_field_spec_input_type_spec.as_mut()
    {
        hook(field_spec);
    }
    match field_spec {
        shared::ast::InputFieldSpec::Literal(literal) => {
            if let Some(hook) =
                hooks.visit_literal_field_spec_input_type_spec.as_mut()
            {
                hook(literal)
            }
            visit_input_type_spec(hooks, &literal.r#type)
        }
        shared::ast::InputFieldSpec::Array(array) => {
            if let Some(hook) =
                hooks.visit_array_field_spec_input_type_spec.as_mut()
            {
                hook(array)
            }
            visit_input_type_spec(hooks, &array.r#type)
        }
    }
}

fn visit_field_definition_input_field_spec(
    hooks: &mut ASTVisitorHooks,
    field: &shared::ast::FieldDefinition<shared::ast::InputFieldSpec>,
) {
    if let Some(hook) = hooks.visit_field_definition_input_field_spec.as_mut() {
        hook(field);
    }
    visit_input_field_spec(hooks, &field.spec)
}

pub fn visit_client_schema(
    hooks: &mut ASTVisitorHooks,
    schema: &client::schema::ClientSchema,
) {
    for directive in schema.directives.values() {
        if let Some(hook) = hooks.visit_client_directive.as_mut() {
            hook(directive);
        }
        if let Some(hook) = hooks.visit_client_directive_location.as_mut() {
            for location in directive.locations.iter() {
                hook(location);
            }
        }
    }
    for fragment in schema.fragments.values() {
        if let Some(hook) = hooks.visit_fragment.as_mut() {
            hook(fragment);
        }
        visit_fragment_spec(hooks, &fragment.borrow().spec);
    }

    for operation in schema.operations.values() {
        if let Some(hook) = hooks.visit_operation.as_mut() {
            hook(operation);
        }
        for parameter in operation.borrow().parameters.values() {
            visit_field_definition_input_field_spec(hooks, parameter);
        }
        visit_fragment_spec(hooks, &operation.borrow().fragment_spec);
    }
}
